cedar-policy-validator/src/schema/raw_name.rs

```diff
@@ -0,0 +1,93 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use cedar_policy_core::ast::{Id, Name};
+use cedar_policy_core::parser::Loc;
+use serde::{Deserialize, Serialize};
+
+/// A newtype which indicates that the contained `Name` may not yet be
+/// fully-qualified.
+///
+/// You can convert it to a fully-qualified `Name` using `.qualify_with()`.
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
+#[serde(transparent)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+pub struct RawName(Name);
+
+impl RawName {
+    /// Create a new `RawName` from the given `Id`
+    pub fn new(id: Id) -> Self {
+        Self(Name::unqualified_name(id))
+    }
+
+    /// Create a new `RawName` from a basename, namespace components as `Id`s, and optional source location
+    pub fn from_components(
+        basename: Id,
+        namespace: impl IntoIterator<Item = Id>,
+        loc: Option<Loc>,
+    ) -> Self {
+        Self(Name::new(basename, namespace, loc))
+    }
+
+    /// Create a new `RawName` by parsing the provided string, which should contain
+    /// an unqualified `Name` (no explicit namespaces)
+    pub fn parse_unqualified_name(
+        s: &str,
+    ) -> Result<Self, cedar_policy_core::parser::err::ParseErrors> {
+        Name::parse_unqualified_name(s).map(RawName)
+    }
+
+    /// Create a new `RawName` by parsing the provided string, which should contain
+    /// a `Name` in normalized form.
+    ///
+    /// (See the [`cedar_policy_core::FromNormalizedStr`] trait.)
+    pub fn from_normalized_str(
+        s: &str,
+    ) -> Result<Self, cedar_policy_core::parser::err::ParseErrors> {
+        use cedar_policy_core::FromNormalizedStr;
+        Name::from_normalized_str(s).map(RawName)
+    }
+
+    /// Is this `RawName` unqualified, that is, written without any _explicit_
+    /// namespaces.
+    /// (This method returning `true` does not imply that the `RawName` will
+    /// _eventually resolve_ to an unqualified name.)
+    pub fn is_unqualified(&self) -> bool {
+        self.0.is_unqualified()
+    }
+
+    /// Convert this `RawName` to a `Name` by adding the given `ns` as its
+    /// prefix, or by no-op if `ns` is `None`.
+    ///
+    /// Note that if the `RawName` already had a non-empty explicit namespace,
+    /// no additional prefixing will be done, even if `ns` is `Some`.
+    pub fn qualify_with(self, ns: Option<&Name>) -> Name {
+        self.0.prefix_namespace_if_unqualified(ns)
```

Comment:
Do we still want this function on `Name`? Feels like it should only exist for `RawName` if possible. Or, if we need both, they should probably use the same name.

Reply:
`Name` is a Core type, and I think DRT relies on that function being available on Core `Name`. We could switch DRT over to this `RawName` type as well at some point, at which time we should probably move `RawName` into Core.  That seems out of scope for this PR in my opinion, which is already a large enough diff.

Which name do you prefer?

Reply:
either is fine

Reply:
Renamed the Core function to `qualify_with` in the latest rev